tracing/tracing.sh

```diff
@@ -31,3 +31,145 @@ function sync_main_kw_file_with_tracing()
   printf '%s' "$main_kw_file_with_tracing" > "${bin}/${kw_path}"
   chmod +x "${bin}/${kw_path}"
 }
+
+# This function injects code that enables tracing in every kw library file (files
+# inside the `src` directory) and installs them. Files pertaining to Bash and Zsh
+# completions (`src/bash_autocomplete.sh` and `src/_kw`), and the `src/VERSION`
+# aren't injected.
+#
+# @input_dir: Path to directory where base library files are stored
+# @output_dir: Path to directory where altered files with tracing will be installed
+function sync_kw_lib_files_with_tracing()
+{
+  local input_dir="$1"
+  local output_dir="$2"
+  local file_with_tracing
+  local parent_directory
+  local all_filepaths
+  local tmp_dir_path
+
+  tmp_dir_path=$(mktemp --directory)
+
+  all_filepaths=$(find "$input_dir" -type f)
+
+  while IFS=$'\n' read -r filepath; do
+    # We shouldn't try to inject code in the completions files nor the `VERSION` file
+    if [[ "$filepath" =~ src\/(_kw|bash_autocomplete\.sh|VERSION) ]]; then
+      cp "$filepath" "${tmp_dir_path}/$(basename "$filepath")"
+    else
+      file_with_tracing=$(inject_lib_file_with_tracing "$filepath")
+      # Removing 'src/' from the filepath
+      filepath=$(printf '%s' "$filepath" | cut --delimiter='/' -f2-)
+      # In case the file is in a sub-directory of `src`, create it
+      parent_directory="${filepath%/*}"
+      if [[ ! "$parent_directory" =~ \.sh$ ]]; then
+        mkdir --parents "${tmp_dir_path}/${parent_directory}"
+      fi
+      # Save altered file in temporary directory
+      printf '%s' "$file_with_tracing" > "${tmp_dir_path}/${filepath}"
+    fi
+  done <<< "$all_filepaths"
+
+  # Commit the synchronization of altered files
+  rsync --quiet --recursive "${tmp_dir_path}/" "${output_dir}"
+}
+
+# This function reads a kw library file, injects it with code that logs the execution
+# for tracing and outputs this altered file.
+#
+# @filepath: Path of file to be altered
+#
+# Return:
+# Outputs the file in `@filepath` injected with code that enables tracing.
+function inject_lib_file_with_tracing()
+{
+  local filepath="$1"
+  local and_return_statement
+  local and_exit_statement
+  local file_with_tracing
+  local inside_function
+  local function_name
+
+  filepath=$(realpath "$filepath")
+
+  and_return_statement='&& return'
+  and_exit_statement='&& exit'
+
+  while read -r line; do
+    # Start of a function declaration and definition
+    if [[ "$line" =~ ^function[[:space:]].+\(\)$ ]]; then
+      inside_function=1
+      function_name=$(printf '%s' "$line" | cut --delimiter=' ' -f2 | sed -e 's/()//')
+      file_with_tracing+="$line"$'\n'
+      file_with_tracing+='{'$'\n'
+      file_with_tracing+="$(get_tracing_log_line 'entry' "$function_name")"$'\n'
+      read -r line
+    else
+      file_with_tracing+="$line"$'\n'
+    fi
+
+    while [[ -n "$inside_function" ]]; do
```

Comment:
Maybe you can create a dedicated function for the operations inside this loop; this can also enable you to add a lot of refined tests for this parser. Also, I think the switch/case will be more readable for this case.

Reply:
Done